File: go/internal/sol/p0076/p0076.go
package p0076

/*
another version of problem 31

numbers: 1, 2, 3, ..., 99
total: 100
*/

import (
	"slices"
	"strconv"
)

func compute(numbers []int, target int) string {
	tbl := make([]int, target+1)
	tbl[0] = 1

	for c := range slices.Values(numbers) {
		for i := c; i <= target; i++ {
			tbl[i] += tbl[i-c]
		}
	}

	return strconv.FormatInt(int64(tbl[len(tbl)-1]), 10)
}

func Solve() string {
	const total = 100
	numbers := make([]int, total-1)
	for i := range len(numbers) {
		numbers[i] = i + 1
	}

	return compute(numbers, total)
}

File: haskell/src/Sol/P0076.hs
module Sol.P0076 (compute, solve) where

{-
another version of problem 31
  coins: 1, 2, 3, ..., 99
  total: 100
-}

import Data.Array.ST (modifyArray, newArray, readArray, runSTUArray, writeArray)
import Data.Array.Unboxed (UArray, (!))
import Data.Foldable (for_)

dpSolve :: Int -> [Int] -> UArray Int Int
dpSolve target coins =
    runSTUArray $ do
        memo <- newArray (0, target) 0
        writeArray memo 0 1
        for_ coins $ \c ->
            for_ [c .. target] $ \i -> do
                tmp <- readArray memo (i - c)
                modifyArray memo i (+ tmp)
        pure memo

compute :: Int -> [Int] -> String
compute target coins =
    show $ dpSolve target coins ! target

solve :: String
solve = compute 100 [1 .. 99]
